### DISCUSSION TITLE
Next 13.4 RFC/Ideas (Server Actions, Tags, revalidate, cache)

### DISCUSSION THREAD

**Discussion by matannahmani** [trpc/trpc#4333] [RFC / Ideas]
Hey, I love using TRPC with NextJS and wanted to share some ideas I've been playing with during the weekend.
I tried to solve some of the new problems that have been raised up with the Next 13 Appdir

I couldn't really think of an easy way to do batch requests with caching yet. 

First, we can create a new wrapper for the vanilla proxy client which allow us to use Next 13 custom fetch functions (tags, revalidate, cache)
```ts
export const next14Client = createTRPCNextProxyClient<AppRouter>(trpcConfigSSR);
// custom behavior
next14Client.v1.anime.list.query(undefined, {
    next: {
        revalidate: 60,
    }
})
// equals to
fetch("v1.anime.list", {
    'next': {
        'revalidate': 60, // custom time
        'tags': ["v1.anime.list"] // default tags
    }
})

// custom behavior with input
next14Client.v1.anime.list.query({
    order: 'ASC'
}, {
    next: {
        revalidate: 60,
    }
})
// equals to
fetch("v1.anime.list", {
    'next': {
        'revalidate': 60, // custom time
        'tags': ["v1.anime.list.order-asc",] // adjusted tag (default tags + input stringified)
    }
})
```
In addition, we can create a new function to retrieve the tags easily to invalidate queries easily on the backend:
```ts
// allows us to have a typesafe way to retrieve tags
const getTag = createTRPCTagsRetriever<AppRouter>();

// default behavior
getTag('v1.anime.list')
// or
getTag('v1.anime.list', {order: 'ASC'})
```

And lastly, we create a custom hook that allows us to easily call a server action (mutation)
```ts
// lastly we might add a custom hook to make server actions easily
const useServerAction = createTRPCServerAction<AppRouter>(trpcConfigSSR);

// client usage | inspired by zact https://github.com/pingdotgg/zact
// component.tsx
"use client";

import { useServerAction } from "~/utils/api";

export const trpcServerComp = () => {
  const { mutate, data, isLoading } = useServerAction("v1.anime.createOne");

  return (
    <div>
      <button onClick={() => mutate({ name: "bleach" })}>
        Create anime
      </button>
      {isLoading && <div>Loading...</div>}
      {data?.message}
    </div>
  );
};
```

**Comment by arashi-dev** [comment]
that sounds great! 
but we should also consider that using "dot case string identifiers" that were used in old trpc api, is something that no one is gonna like as it loses the type-reference and makes the debugging much harder. I'd like it to be like this `trpc.admin.posts.create.useServerAction` like we just add a `useServerAction` hook beside `useMutation` . I'm not sure about side effects or even if it is possible, but as a user who uses trpc, this is the simplest api I expect it to be

**Reply by matannahmani** [reply]
After playing around with it a little bit more, I tend to think that something like [createTRPCNext ](https://trpc.io/docs/nextjs/setup) might not be needed and just the vanilla client is good enough, but we need more time to experiment and play with Next 13.
It will probably take a few more weeks till we can get a better grasp of the situation and structure a good draft.

**Reply by arashi-dev** [reply]
@matannahmani 
I'm agreed with needing more time to experiment and play with Next 13 as its APIs are new and the new mutation api is in the alpha stage, so we may face a lot of bugs and changes in the future.

have you seen the [zact](https://github.com/pingdotgg/zact) library which was implmeneted by theo?
we can have inspiration from its code and solutions. However, I'm guessing he is already working it around with the trpc maintainers 

it would be great if trpc maintainers share their thoughts about their plans.
I have not much knowledge about trpc codebase, but I would like to have contribute if they need help
